File: go/models/accommodation.go
package models

import (
	"challenge-flutter-go/api/utils/geo"
	"errors"
	"time"

	"gorm.io/gorm"
)

type AccommodationType string

const (
	AccommodationTypeHotel  AccommodationType = "hotel"
	AccommodationTypeAirbnb AccommodationType = "airbnb"
	AccommodationTypeOther  AccommodationType = "other"
)

type Accommodation struct {
	gorm.Model
	AccommodationType AccommodationType `gorm:"not null; default:'hotel'"`
	StartDate         time.Time         `gorm:"not null"`
	TripID            uint
	Trip              Trip      `gorm:"constraint:OnUpdate:CASCADE"`
	EndDate           time.Time `gorm:"not null"`
	Address           string    `gorm:"not null"`
	BookingURL        string    `gorm:"null"`
	Name              string    `gorm:"not null"`
	Latitude          float64   `gorm:"default:0"`
	Longitude         float64   `gorm:"default:0"`
	Price             float64   `gorm:"not null; default:0" validate:"min=0"`
}

func (a *Accommodation) isAccommodationTypeValid() (isValid bool) {
	return a.AccommodationType == AccommodationTypeHotel || a.AccommodationType == AccommodationTypeAirbnb || a.AccommodationType == AccommodationTypeOther
}

func (a *Accommodation) BeforeSave(tx *gorm.DB) (err error) {
	if !a.isAccommodationTypeValid() {
		return gorm.ErrInvalidData
	}
	if err = geo.GetGeoLocation(a.Address, &a.Latitude, &a.Longitude); err != nil {
		return errors.New("geo location api is on an error state")
	}
	return
}

File: move_together_app/lib/core/models/accommodation.dart
import 'package:move_together_app/utils/dynamic_to_double.dart';

enum AccommodationType {
  hotel,
  airbnb,
  other,
}

class Accommodation {
  final int id;
  final AccommodationType accommodationType;
  final DateTime startDate;
  final DateTime endDate;
  final String address;
  final String name;
  final String? bookingUrl;
  final double latitude;
  final double longitude;
  final double price;

  Accommodation({
    required this.id,
    required this.accommodationType,
    required this.startDate,
    required this.endDate,
    required this.address,
    required this.name,
    this.bookingUrl,
    required this.latitude,
    required this.longitude,
    required this.price,
  });

  factory Accommodation.fromJson(Map<String, dynamic> json) {
    return Accommodation(
      id: json['id'],
      accommodationType: AccommodationType.values.firstWhere(
          (e) => e.toString().split('.').last == json['accommodationType']),
      startDate: DateTime.parse(json['startDate']),
      endDate: DateTime.parse(json['endDate']),
      address: json['address'],
      name: json['name'],
      bookingUrl: json['bookingURL'],
      latitude: dynamicToDouble(json['latitude']),
      longitude: dynamicToDouble(json['longitude']),
      price: dynamicToDouble(json['price']),
    );
  }
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'accommodationType': accommodationType.toString(),
      'startDate': startDate,
      'endDate': endDate,
      'address': address,
      'name': name,
      'bookingUrl': bookingUrl,
      'latitude': latitude,
      'longitude': longitude,
      'price': price,
    };
  }

  bool get hasValidGeolocation => latitude != 0 && longitude != 0;
}
